akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphMergePrioritizedSpec.scala

```diff
@@ -0,0 +1,144 @@
+
+      threeSourceMerge(source1, source2, source3, priorities, probe).run()
+
+      val subscription = probe.expectSubscription()
+
+      var collected = Seq.empty[Int]
+      for (_ ← 1 to elementCount) {
+        subscription.request(1)
+        collected :+= probe.expectNext()
+      }
+
+      val ones = collected.count(_ == 1).toDouble
+      val twos = collected.count(_ == 2).toDouble
+      val threes = collected.count(_ == 3).toDouble
+
+      (ones / twos).round shouldEqual 2
+      (ones / threes).round shouldEqual 6
+      (twos / threes).round shouldEqual 3
+    }
+
+    "stream data when only one source produces" in {
+      val elementCount = 10
+      val source1 = Source.fromIterator(() ⇒ Seq.fill(elementCount)(1).iterator)
+      val source2 = Source.fromIterator(() ⇒ Seq.empty[Int].iterator)
+      val source3 = Source.fromIterator(() ⇒ Seq.empty[Int].iterator)
+
+      val priorities = Seq(6, 3, 1)
+
+      val probe = TestSubscriber.manualProbe[Int]()
+
+      threeSourceMerge(source1, source2, source3, priorities, probe).run()
+
+      val subscription = probe.expectSubscription()
+
+      var collected = Seq.empty[Int]
+      for (_ ← 1 to elementCount) {
+        subscription.request(1)
+        collected :+= probe.expectNext()
+      }
+
+      val ones = collected.count(_ == 1)
+      val twos = collected.count(_ == 2)
+      val threes = collected.count(_ == 3)
+
+      ones shouldEqual elementCount
+      twos shouldEqual 0
+      threes shouldEqual 0
+    }
+
+    "stream data with priority when only two sources produce" in {
+      val elementCount = 10000
+      val source1 = Source.fromIterator(() ⇒ Seq.fill(elementCount)(1).iterator)
+      val source2 = Source.fromIterator(() ⇒ Seq.fill(elementCount)(2).iterator)
+      val source3 = Source.fromIterator(() ⇒ Seq.empty[Int].iterator)
+
+      val priorities = Seq(6, 3, 1)
+
+      val probe = TestSubscriber.manualProbe[Int]()
+
+      threeSourceMerge(source1, source2, source3, priorities, probe).run()
+
+      val subscription = probe.expectSubscription()
+
+      var collected = Seq.empty[Int]
+      for (_ ← 1 to elementCount) {
+        subscription.request(1)
+        collected :+= probe.expectNext()
+      }
+
+      val ones = collected.count(_ == 1).toDouble
+      val twos = collected.count(_ == 2).toDouble
+      val threes = collected.count(_ == 3)
+
+      threes shouldEqual 0
+      (ones / twos).round shouldBe 2
+    }
+  }
+
+  private def threeSourceMerge[T](source1: Source[T, NotUsed], source2: Source[T, NotUsed], source3: Source[T, NotUsed], priorities: Seq[Int], probe: ManualProbe[T]) = {
+    RunnableGraph.fromGraph(GraphDSL.create(source1, source2, source3)((_, _, _)) { implicit b ⇒ (s1, s2, s3) ⇒
+      val merge = b.add(MergePrioritized[T](priorities))
+      s1.out ~> merge.in(0)
+      s2.out ~> merge.in(1)
+      s3.out ~> merge.in(2)
+      merge.out ~> Sink.fromSubscriber(probe)
+      ClosedShape
+    })
+  }
+}
```

Comment:
I'd like to see some test coverage for upstream and downstream failures as well, I think there are no bugs in those parts of the stage but that test coverage would still be good make sure.